Drawing Tool Improvements

JakobMiksch opened the pull request:
- references #564 
- snapped node or edges are highlighted
- provides various configurable styles for editing
- todo: 
  - [x] handle polygons

## snapping

![Peek 2022-07-27 18-55](https://user-images.githubusercontent.com/15704517/181305357-0eb63d86-0945-4506-ab9d-2696c13bd548.gif)

## highlight vertices of snapped line

![Peek 2022-07-28 10-25](https://user-images.githubusercontent.com/15704517/181458542-c5320424-2dee-4b52-9767-0f8a3779bcb9.gif)

JakobMiksch commented:
@geographika this PR so far provides a default style. I guess there is a need to manually adjust the Style via the configs in `CpsiMapview.view.tool.DrawingButton`

possible options would be:
- start/end point
  - color
  - shape
  - size
- line 
  - color
  - width

shall we expose these properties as config? Which one would be relevant here?

geographika commented:
@JakobMiksch can we do similar to https://github.com/compassinformatics/cpsi-mapview/blob/f12f00982031f167c74c9316cc10c01ff73e7ab2/app/view/button/DigitizeButton.js#L54 and allow a full style object to be passed in as a config?

JakobMiksch commented:
@geographika I updated the PR. I added the configs for the styles in the way you suggested. 
Also I created a user feedback if the edges or node layers are snapped.
Could you tell me if this is the editing behavior you are expecting?

geographika commented:
@JakobMiksch - these updates are great - thanks!
The symbol display when hovering over a snapped feature works very well. 

If I wanted to keep the same start and end symbology once the line is created, would I need to apply the style to the Edges vector layer?

Style when drawing:

![image](https://user-images.githubusercontent.com/490840/181455009-e8825056-550a-4e8b-a392-417aaf6cf98e.png)

Style when completed:

![image](https://user-images.githubusercontent.com/490840/181455045-45d2330b-b0a4-4662-91bf-e3e1345b15e7.png)

JakobMiksch commented:
@geographika I added highlighting for the vertices of the snapped edge. See the new gif in this PR's description. The code is pushed. What is your impression?
(I saw your comment above already, and will check it now)

geographika commented:
Very nice! Showing the vertices will be useful in a number of scenarios. Could we make it configurable via the view on whether vertices should be shown? The tool will be reused in a few places, so being able to turn off vertices would be useful.

geographika commented on app/controller/button/DrawingButtonController.js:
```diff
@@ -44,13 +44,19 @@ Ext.define('CpsiMapview.controller.button.DrawingButtonController', {
     snapInteraction: null,
 
+    /**
+     * If user has started to edit a line i.d. the first point of a line is already set
```
Not sure what i.d. here means?

JakobMiksch replied:
I actually meant `i.e.`, but I simply replaced it with "this means". This might be better understandable.
Fixed with [8a1a771](https://github.com/compassinformatics/cpsi-mapview/pull/578/commits/8a1a7711b7d95d6f15d6ab4b2f9a073733558efd)

geographika replied:
I tried to google the equivalent of e.g. in German to see if it was that, should have thought of i.e. first!

geographika commented on app/controller/button/DrawingButtonController.js:
```diff
@@ -97,19 +103,126 @@ Ext.define('CpsiMapview.controller.button.DrawingButtonController', {
         };
 
         me.drawInteraction = new ol.interaction.Draw(drawInteractionConfig);
+        me.drawInteraction.on('drawstart',me.handleDrawStart);
```
Needs a space after comma to pass linting?

JakobMiksch replied:
fixed with [abc05db](https://github.com/compassinformatics/cpsi-mapview/pull/578/commits/abc05dba3e376f274ce2ecd1c8be494da91234b0)

geographika reviewed (commented):
Only very minor comments

JakobMiksch commented:
> @JakobMiksch - these updates are great - thanks! The symbol display when hovering over a snapped feature works very well.
> 
> If I wanted to keep the same start and end symbology once the line is created, would I need to apply the style to the Edges vector layer?
> 
> Style when drawing:
> 
> ![image](https://user-images.githubusercontent.com/490840/181455009-e8825056-550a-4e8b-a392-417aaf6cf98e.png)
> 
> Style when completed:
> 
> ![image](https://user-images.githubusercontent.com/490840/181455045-45d2330b-b0a4-4662-91bf-e3e1345b15e7.png)

@geographika in https://github.com/compassinformatics/cpsi-mapview/pull/578/commits/76e6a632845578a21c3d36ba1a3c375b027b2907 I added the drawStyle as the default style of the layer. Should this be the default? And should it be possible to configure the default style of the drawn layer, once it is not in editing mode anymore?

geographika commented:
> @geographika in [76e6a63](https://github.com/compassinformatics/cpsi-mapview/commit/76e6a632845578a21c3d36ba1a3c375b027b2907) I added the drawStyle as the default style of the layer. Should this be the default? And should it be possible to configure the default style of the drawn layer, once it is not in editing mode anymore?

I'm not sure if I fully understand the quests here, but I've tested the last update, and this approach is perfect. Keeping the Start and End styles of the line once created is the main requirement (so a user know which end is which). 

If we could make hiding/showing vertices configurable, I think this update is then ready to merge?

JakobMiksch commented:
@geographika do we need to test/implement the snapping for polygon layers as well?

geographika commented:
> @geographika do we need to test/implement the snapping for polygon layers as well?

Ideally, yes, but the demo currently doesn't have a polygon layer to test with. Would the approach be very different from polygons compared to lines?
Are you able to add in the code without testing? I could then test the branch in a local project with polygons. Otherwise, I can add a sample polygon layer to the demo.

JakobMiksch commented:
I assume, polygons would be different from lines, because we might need to check if the cursor is on the boundary or the inside of the polygon.

JakobMiksch commented:
In 06fb051 I added minimal support for polygons (without testing). Maybe you could try if this is already okay for you? 
Otherwise I would need to refine it.

geographika commented:
@JakobMiksch - I've tested the branch locally with a polygon layer and it highlights the snap correctly:

![image](https://user-images.githubusercontent.com/490840/181487832-61c9cc72-2947-424f-a81a-232a1d7d637f.png)

One minor issue is that the vertices are not visible above the polygon layer - does it need to be moved to the top of the rendering stack?

When polygon layer is on:

![image](https://user-images.githubusercontent.com/490840/181487973-8c1a1148-cbcd-4deb-a5fc-2340570c5d55.png)

When polygon layer is off (the white points are now visible):

![image](https://user-images.githubusercontent.com/490840/181488001-bc5943e5-e8ff-4282-8938-9835f35ddfaa.png)

JakobMiksch commented:
@geographika It indeed looks like the highlighted vertices are rendered too low.
Could you try adding this line: `snappedEdgeVertexStyle.setZIndex(Infinity);` in line 208 of `app/controller/button/DrawingButtonController.js`?
This might be an easy fix

jansule commented on app/view/button/DrawingButton.js:
```diff
@@ -60,7 +60,150 @@ Ext.define('CpsiMapview.view.tool.DrawingButton', {
+         * Style of the editing cursor, before first point is set.
+         */
+        drawBeforeEditingPoint: new ol.style.Style({
+            image: new ol.style.Circle({
```
Minor: Styles could be refactored into a dedicated class, in order to ensure better readability of the `DrawingButton` class

JakobMiksch replied:
Good idea. I did it with https://github.com/compassinformatics/cpsi-mapview/pull/578/commits/78769d6a4f510b9ffe8b29cbeef59e3795a98f8d

jansule reviewed (approved):
Nice work @JakobMiksch. I'm wondering if it makes sense to provide this functionality in BasiGX, since it might be useful for others as well.

JakobMiksch commented:
Thanks @jansule thanks for the review. I moved the styles to a util.
Moving the code to BasiGX would be fine for me. But first I would commit the code to this repo.

JakobMiksch commented:
@geographika I moved to styles to a util and made sure they are always created as new instance. My suspicion is that the bug with the vertices rendering might have to do with that. Maybe you can try again now. Otherwise I think I would be best if I had a polygon layer to test for my self.

geographika commented:
> @geographika I moved to styles to a util and made sure they are always created as new instance. My suspicion is that the bug with the vertices rendering might have to do with that. Maybe you can try again now. Otherwise I think I would be best if I had a polygon layer to test for my self.

@JakobMiksch - I've added a polygon layer to the example in https://github.com/compassinformatics/cpsi-mapview/pull/579
It is a minor issue, but if you hover over a line within a polygon the vertices don't show:
![CpsiMapview](https://user-images.githubusercontent.com/490840/181714579-5a34ffe7-e1fd-4cf5-94af-e6fa83893729.gif)

JakobMiksch commented:
Thanks  @geographika for providing the polygon layer. It helped me to find the problem. In the situation of the described bug, not the line feature was returned anymore but the polygon feature. This is fixed now.
I also added a new style ( a black star) indicating when the cursor is snapping or above the referenced polygon layer.

geographika commented:
Tested locally and it works very well with all layer types. If you are happy @JakobMiksch I can go ahead and merge this.

JakobMiksch commented:
@geographika Great to hear that it works! This PR can be merged from my side.
